flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.iceberg.flink.source.ScanContext;
+import org.apache.iceberg.flink.source.StreamingStartingStrategy;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Internal
+public class ContinuousSplitPlannerImpl implements ContinuousSplitPlanner {
+  private static final Logger LOG = LoggerFactory.getLogger(ContinuousSplitPlannerImpl.class);
+
+  private final Table table;
+  private final ScanContext scanContext;
+  private final ExecutorService workerPool;
+
+  public ContinuousSplitPlannerImpl(Table table, ScanContext scanContext) {
+    this.table = table;
+    this.scanContext = scanContext;
+    // Within a JVM, table name should be unique across sources.
+    // Hence it is used as worker pool thread name prefix.
+    this.workerPool = ThreadPools.newWorkerPool("iceberg-worker-pool-" + table.name(), scanContext.planParallelism());
+  }
+
+  @Override
+  public ContinuousEnumerationResult planSplits(IcebergEnumeratorPosition lastPosition) {
+    table.refresh();
+    if (lastPosition != null) {
+      return discoverDeltaSplits(lastPosition);
+    } else {
+      return discoverInitialSplits();
+    }
+  }
+
+  /**
+   * Discover delta changes between @{code lastPosition} and current table snapshot
+   */
+  private ContinuousEnumerationResult discoverDeltaSplits(IcebergEnumeratorPosition lastPosition) {
+    // incremental discovery mode
+    Snapshot currentSnapshot = table.currentSnapshot();
+    if (currentSnapshot.snapshotId() == lastPosition.endSnapshotId()) {
+      LOG.info("Current table snapshot is already enumerated: {}", currentSnapshot.snapshotId());
+      return new ContinuousEnumerationResult(Collections.emptyList(), lastPosition);
+    } else {
+      ScanContext incrementalScan = scanContext
+          .copyWithAppendsBetween(lastPosition.endSnapshotId(), currentSnapshot.snapshotId());
+      LOG.info("Incremental scan: startSnapshotId = {}, endSnapshotId = {}",
+          incrementalScan.startSnapshotId(), incrementalScan.endSnapshotId());
+      List<IcebergSourceSplit> splits = FlinkSplitPlanner.planIcebergSourceSplits(table, incrementalScan, workerPool);
+      IcebergEnumeratorPosition position = IcebergEnumeratorPosition.builder()
+          .startSnapshotId(lastPosition.endSnapshotId())
+          .startSnapshotTimestampMs(lastPosition.endSnapshotTimestampMs())
+          .endSnapshotId(currentSnapshot.snapshotId())
+          .endSnapshotTimestampMs(currentSnapshot.timestampMillis())
+          .build();
+      LOG.info("Discovered {} splits from incremental scan: {}", splits.size(), position);
+      return new ContinuousEnumerationResult(splits, position);
+    }
+  }
+
+  /**
+   * Discovery initial set of splits based on {@link StreamingStartingStrategy}
+   */
+  private ContinuousEnumerationResult discoverInitialSplits() {
```

Comment:
Can you add to the javadoc that this only returns splits if there is a starting table scan?

Reply:
And that this is going to return a starting position of the starting snapshot, which is exclusive.

Reply:
will update